### PR TITLE
Feature/add tokenfactory

### PR DESCRIPTION
**Pull request by matpb**
Adding Osmosis TokenFactory module.
Taken from https://github.com/CosmWasm/token-factory

Extra work done:
- new way to generate docs (./scripts/generate_docs.sh)
- added StoreLoader upgrade
- **upgrade required**, changes done to ./app/upgrade.go

### CODE REVIEW

File: app/upgrade.go [hunk 1]
```diff
@@ -16,8 +17,20 @@ func NewUpgradeHandler(app *WasmApp) UpgradeHandler {
 	return UpgradeHandler{app}
 }
 
-func (h UpgradeHandler) CreateUpgradeHandler() upgradetypes.UpgradeHandler {
+func (h UpgradeHandler) CreateUpgradeHandler(app *WasmApp) upgradetypes.UpgradeHandler {
 	return func(ctx sdk.Context, _ upgradetypes.Plan, vm module.VersionMap) (module.VersionMap, error) {
-		return h.mm.RunMigrations(ctx, h.configurator, vm)
+		logger := ctx.Logger().With("upgrade", UpgradeName)
```
Comment (jumpdest7d) [inline app/upgrade.go:22]: love this context!!

File: scripts/generate_docs.sh [hunk 2]
```diff
@@ -0,0 +1,56 @@
+ #!/usr/bin/env bash
+
+set -eo pipefail
+
+mkdir -p ./tmp-swagger-gen
+
+# move the vendor folder to a temp dir so that go list works properly
+temp_dir="f29ea6aa861dc4b083e8e48f67cce"
```
Comment (jumpdest7d) [inline scripts/generate_docs.sh:8]: What about using `mktemp -d` - this creates a unique temporary directory and ensures there are no collisions. This also makes the script re-runnable by doing away with having to check if the temp directory already exists on subsequent runs.

I found this stackoverflow post about it: https://stackoverflow.com/a/4632032

File: scripts/generate_docs.sh [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+if [ -d $temp_dir ]; then
+  mv ./$temp_dir ./vendor
+fi
+
+proto_dirs=$(find ./proto "$cosmos_sdk_dir"/proto "$tokenfactory_dir"/proto -path -prune -o -name '*.proto' -print0 | xargs -0 -n1 dirname | sort | uniq)
+for dir in $proto_dirs; do
+  # generate swagger files (filter query files)
+  query_file=$(find "${dir}" -maxdepth 1 \( -name 'query.proto' -o -name 'service.proto' \))
+  if [[ ! -z "$query_file" ]]; then
+    protoc \
```
Comment (jumpdest7d) [inline scripts/generate_docs.sh:29]: should we do a prerequisite check for `protoc` at the top of the script? Might be nice to tell users to install this before the script runs.

File: scripts/generate_docs.sh [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      --swagger_opt simple_operation_ids=true
+  fi
+done
+
+cd ./docs
+yarn install
+yarn combine
+
+#Add public servers to spec file for Noria testnet and mainnet
+yq -i '."host"="archive-lcd.noria.nextnet.zone"' static/swagger.yaml
```
Comment (jumpdest7d) [inline scripts/generate_docs.sh:47]: we should set the archive node as a variable right at the top of the script

**Review by jumpdest7d — APPROVED** [review]
A few suggestions @matpb, but overall looks GTG!

File: scripts/generate_docs.sh (hunk 2 above)
Comment (matpb) [reply scripts/generate_docs.sh]: The `tmp-swagger-gen` needs to be hardcoded for `./docs/config.json` to work properly.
I do like that mktemp approach for the vendor hack though, thank you!

Fixed in https://github.com/noria-net/noria/pull/11/commits/98aca435addf76ada0ac6f3e361c028907d734f1

File: scripts/generate_docs.sh (hunk 3 above)
Comment (matpb) [reply scripts/generate_docs.sh:36]: Good call, fixed in baf772a4cbcaf8a153577a3b05852a3ea8935a12

File: scripts/generate_docs.sh (hunk 4 above)
Comment (matpb) [reply scripts/generate_docs.sh]: good idea, fixed in aa08b0a3ca5686fe61c58824905255284b521f05

**Review by jumpdest7d — APPROVED** [review]
Looking superb @matpb!

**Review by jumpdest7d — APPROVED** [review]

**Pull request merged**